File: leetcode/clone_graph/CloneGraph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (node == NULL) return NULL; 
        
        map<UndirectedGraphNode *, UndirectedGraphNode *> mapping; 
        set<UndirectedGraphNode *> history; 
        helper(node, mapping, history);
        
        return mapping[node]; 
    }
    
    void helper(UndirectedGraphNode *node, map<UndirectedGraphNode *, UndirectedGraphNode*> &mapping, set<UndirectedGraphNode *> &history) {
        if (history.find(node) != history.end()) return; 
        
        if (mapping.find(node) == mapping.end()) {
            UndirectedGraphNode * newNode = new UndirectedGraphNode(node->label); 
            mapping[node] = newNode; 
        }
        
        for (int i = 0; i < node->neighbors.size(); i++) {
            if (mapping.find(node->neighbors[i]) == mapping.end()) {
                UndirectedGraphNode * newNeighbor = new UndirectedGraphNode(node->neighbors[i]->label); 
                mapping[node->neighbors[i]] = newNeighbor; 
            }
            
            mapping[node]->neighbors.push_back(mapping[node->neighbors[i]]); 
        }
        
        history.insert(history.begin(), node); 
        
        for (int i = 0; i < node->neighbors.size(); i++) {
            helper(node->neighbors[i], mapping, history); 
        }
    }
};

File: lintcode/clone_graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     ArrayList<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        Map<UndirectedGraphNode, UndirectedGraphNode> m = new HashMap<>(); 
        Set<UndirectedGraphNode> history = new HashSet<>(); 
        Queue<UndirectedGraphNode> queue = new LinkedList<>(); 
        
        if (node != null) {
            queue.add(node);
        }
        
        while (queue.size() != 0) {
            UndirectedGraphNode curr = queue.remove(); 
            
            if (history.contains(curr)) {
                continue; 
            }
            
            if (!m.containsKey(curr)) {
                UndirectedGraphNode temp = new UndirectedGraphNode(curr.label);
                m.put(curr, temp);
            }
            
            for (UndirectedGraphNode child : curr.neighbors) {
                if (!m.containsKey(child)) {
                    UndirectedGraphNode temp = new UndirectedGraphNode(child.label);
                    m.put(child, temp);
                }
                
                m.get(curr).neighbors.add(m.get(child));
                queue.add(child);
            }
            
            history.add(curr);
        }
        
        return m.get(node);
    }
}
